yapapi/net.py

```diff
@@ -0,0 +1,156 @@
+        net_api: "Net",
+        ip: str,
+        owner_id: str,
+        owner_ip: Optional[str] = None,
+        network_id: Optional[str] = None,
+        mask: Optional[str] = None,
+        gateway: Optional[str] = None,
+    ):
+        self._net_api = net_api
+        network_ip = f"{ip}/{mask}" if mask else ip
+        try:
+            self._ip_network: IpNetwork = ip_network(network_ip, strict=False)
+        except ValueError as e:
+            raise NetworkError(f"{e}.")
+        self._hosts = self._ip_network.hosts()
+
+        self._network_id = network_id or "0x" + uuid4().hex
+        self._gateway = gateway
+        self._owner_id = owner_id
+        self._owner_ip = owner_ip or self._next_address()
+        self._nodes: Dict[str, Node] = dict()
+
+    def __str__(self) -> str:
+        return f"""Network {{
+        id: {self._network_id}
+        ip: {self.network_address}
+        mask: {self.netmask}
+        owner_ip: {self._owner_ip}
+        nodes: {self.nodes_dict}
+    }}"""
+
+    @property
+    def owner_ip(self) -> str:
+        return str(self._owner_ip)
+
+    @property
+    def network_address(self) -> str:
+        return str(self._ip_network.network_address)
+
+    @property
+    def netmask(self) -> str:
+        return str(self._ip_network.netmask)
+
+    @property
+    def gateway(self) -> Optional[str]:
+        if self._gateway:
+            return str(self._gateway)
+        return None
+
+    @property
+    def nodes_dict(self) -> Dict[str, str]:
+        return {str(v.ip): k for k, v in self._nodes.items()}
+
+    @property
+    def network_id(self):
+        return self._network_id
+
+    def _ensure_ip_in_network(self, ip: str):
+        if ip_address(ip) not in self._ip_network:
+            raise NetworkError(
+                f"The given IP ('{ip}') address must belong to the network ('{self._ip_network.with_netmask}')."
+            )
+
+    async def add_address(self, ip: str):
+        """Assign the given IP address to the requestor in the Network."""
+        self._ensure_ip_in_network(ip)
+
+        if ip in self.nodes_dict.keys():
+            raise NetworkError(f"'{ip}' has already been assigned in this network.")
+
+        self._nodes[self._owner_id] = Node(network=self, node_id=self._owner_id, ip=ip)
+
+        await self._net_api.add_address(self, ip)
+
+    async def add_node(self, node_id: str, ip: Optional[str] = None) -> Node:
+        """Add a new node to the network."""
+        if ip:
+            self._ensure_ip_in_network(ip)
+            if ip in self.nodes_dict.keys():
+                raise NetworkError(f"'{ip}' has already been assigned in this network.")
+        else:
+            while True:
+                ip = str(self._next_address())
+                if ip not in self.nodes_dict.keys():
+                    break
+
+        await self._net_api.add_node(self, node_id, ip)
+
+        node = Node(network=self, node_id=node_id, ip=ip)
+        self._nodes[node_id] = node
```

Comment:
I'm not sure, but I think we could have a race condition here?

IP is added to a list of "used" ips here, in line 145.
If we have two calls to `network.add_node` running at the same time with **the same IP** they could pass the validation in line 134 (`if "ip" in self.nodes_dict.keys()`) and both get to `await self._net_api.add_node(...)` and I don't think we want this.

Reply:
addressed this with a lock